optimizer/resolver.go

```diff
@@ -729,3 +742,20 @@ func (nr *nameResolver) handlePosition(pos *ast.PositionExpr) {
 		}
 	}
 }
+
+func (nr *nameResolver) handleUnionSelectList(u *ast.UnionSelectList) {
+	firstSelFields := u.Selects[0].GetResultFields()
+	unionFields := make([]*ast.ResultField, len(firstSelFields))
+	// Copy first result fields, because we may change the result field type.
+	for i, v := range firstSelFields {
```

Comment:
In `resolver.go`, we create result fields for order by clause to resolve to, in `buildUnion`, since we have inferred type already,  we can update the union result field type.